### DISCUSSION TITLE
What is the purpose of the ansible linter's ANSIBLE_DIRECTORY variable?

### DISCUSSION THREAD

**Discussion by lcts** [oxsecurity/megalinter#2620] [Q&A] [answered]
Hi,
 I had assumed that the config variable ANSIBLE_DIRECTORY of [the ansible_lint linter](https://megalinter.io/latest/descriptors/ansible_ansible_lint/) would set the directory ansible_lint lints, i.e it is called with `ansible-lint <options> <ANSIBLE_DIRECTORY>` but instead the linter always lints the entire project.

So, what is the variable *actually* used for?

**Comment by Kurt-von-Laven** [comment]
By the usual naming convention, the config variable would be named `ANSIBLE_ANSIBLE_LINT_DIRECTORY`, so it could be a documentation bug?

**Reply by lcts** [reply]
I don't think it's that.
It works in enabling/disabling the linter (the documentation states that the linter only runs if the directory exists): Depending on what I set the variable to it either runs or doesn't, so the var is used for that at least. But that purpose is already served by the usual enable/disable vars.
I though that maybe ANSIBLE_DIRECTORY would be added to ansible's search path for roles/collections/etc, but it doesn't seem to do that either.

**Reply by Kurt-von-Laven** [reply]
It sounds like you are assuming that it behaves consistently and correctly, which seems unlikely given what you reported. Not at my desk right now, but you could also try `git grep ANSIBLE_DIRECTORY` if you haven't already. I have found GitHub's web search unreliable.

**Reply by lcts** [reply]
hm, you might be right. `megalinter/descriptors/schemas/megalinter-configuration.jsonschema.json` lists both a `ANSIBLE_ANSIBLE_LINT_DIRECTORY` and the `ANSIBLE_DIRECTORY` mentioned in the documentation. I'll give the former a try tomorrow, see if that behaves as expected.

**Reply by Kurt-von-Laven** [reply, marked as answer]
You might also try passing `ANSIBLE_ANSIBLE_LINT_ARGUMENTS: --project-dir <path>`. According to [ansible-lint's documentation](https://ansible-lint.readthedocs.io/usage/), `--project-dir` specifies the:

> Location of project/repository, autodetected based on location of configuration file.

**Reply by lcts** [reply]
Yeah, that works. Since ansible-lint's syntax is `ansible_lint <options> <lintables>` one can also just pass the directory as an option without `--project-dir`

I tested ANSIBLE_ANSIBLE_LINT_DIRECTORY, that variable seems to be unused.

**Comment by nvuillam** [comment]
Some linters indeed works only if a directory is found, that's the case with ansible-lint, and you can override the name of such directory
But the linter remains run on the repo root level, like all the linters :)

**Reply by lcts** [reply]
Yeah, but does it do anything beyond that? If it only enables the linter then it's just a very unintuitive way of putting ANSIBLE_ANSIBLE_LINT in ENABLE_LINTERS, no? Sure, it provides some sort of "autodetection", but there is no reason to expect ansible files to be located in a directory named 'ansible', and if I have to set the dir manually anyways, well, then I can also just manually enable the linter via the normal route.

**Reply by nvuillam** [reply]
This is indeed autodetection that ansible-lint must be run :)
I'm no expert at all, maybe you know other finer ways to detect that a repo contains ansible sources ?

**Reply by lcts** [reply]
I don't think there are any beyond whatever heuristic ansible_lint itself uses - ansible is a bit annoying as it's really just a bunch of yaml files that look a lot like other yaml files. 

But if megalinter called ansible_lint with `ansible_lint $ANSIBLE_ANSIBLE_LINT_ARGUMENTS $ANSIBLE_DIRECTORY` instead of just `ansible_lint $ANSIBLE_ANSIBLE_LINT_ARGUMENTS` it would behave exactly as one would expect and as the documentation suggests it does. Or alternatively just disable it by default and let the user enable it for repos with ansible stuff in it via the normal way, instead of via setting a directory that is not actually used for anything.

It's not that it doesn't work, it's just that it works confusingly.
Anyways, ANSIBLE_ANSIBLE_LINT_ARGUMENTS="--project-dir ${ANSIBLE_DIRECTORY}" as @Kurt-von-Laven suggested can be used as a workaround.

Great work on megalinter, btw. I've switched all my linting over to it and it's an absolute joy.
